src/Components/Components/src/PersistentComponentState.cs

```diff
@@ -46,9 +55,43 @@ public PersistingComponentStateSubscription RegisterOnPersisting(Func<Task> call
     {
         ArgumentNullException.ThrowIfNull(callback);
 
-        _registeredCallbacks.Add(callback);
+        if (callback.Target is not IComponent)
+        {
+            throw new InvalidOperationException("Cannot infer serialization mode for non component. Provide a serialization mode.");
+        }
+
+        var serializationMode = SerializationMode;
```

Comment:
When the code is executed on the circuit (or webassembly runtime) there is no way to get the render mode of the component. For this reason I changed [ComponentStatePersistenceManager.RestoreStateAsync](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-0386c940190e808072229a79b79b845e646152920641a931b44e619ceb4f1efcR51) to provide the serialization mode.

Reply:
Thanks for the suggestion! I tried it out and it won't work. 
The reason is that `ComponentStatePersistenceManager.PersistStateAsync(store, serializationMode, dispatcher)` persists the state for the provided serializationMode (`Server` or `WebAssembly`) into the provided store. We want to store server state and webassembly state into differrent stores. Currently we store the state in `ProtectedPrerenderComponentApplicationStore` for Blazor Server and `PrerenderComponentApplicationStore` for Blazor WebAssembly.

That is why it seemed like a good idea at the time to change `ComponentStatePersistenceManager.RestoreStateAsync(store, serializationMode)` by adding serializationMode argument. We persist the state for the given serialization mode in the given store and similarly restoring the the state for serialization mode from the given store.
But I agree that in the implementation `RestoreStateAsync` doesn't do what you'd expect and indeed ,as you said, it relies on the fact that RestoreStateAsync always gets called before RegisterOnPersisting.

I found a way to fix it. I added `GlobalSerializationMode` in [ISerializationModeHandler](https://github.com/dotnet/aspnetcore/pull/50373/files#diff-b58ef65f1a58ba6b1c30959855cf80b9d8d935681fdc98ed28cb719086b338e6R21). I change the `GlobalSerializationMode` before I call `RestoreStateAsync`. 
`GlobalSerializationMode` is also useful in the case when the code is running on the circuit(or webassembly runtime) and you try to register a callback on webassembly(server). The callback should not be fired. By checking the `GlobalSerializationMode` I throw the exception.

Reply:
> Thanks for the suggestion! I tried it out and it won't work.
> The reason is that ComponentStatePersistenceManager.PersistStateAsync(store, serializationMode, dispatcher) persists the state for the provided serializationMode (Server or WebAssembly) into the provided store. We want to store server state and webassembly state into differrent stores. 

Yeah, that makes sense - it looks like I didn't cover that detail in my initial comment here. I sketched out some of the ideas from my first comment [as a commit](https://github.com/dotnet/aspnetcore/commit/f7a8552d1de5076e123f48f7cf06f8a4224ac33e), because code is probably a lot easier to parse than a list of bullet points 😄

What do you think? The goal with the ideas I'm proposing is to reduce the amount of code dedicated to tracking Server/WebAssembly-specific state. Feel free to borrow some of those ideas if you agree with them (or ignore them if you prefer what you have).

Your approach using a `GlobalSerializationMode` property is also valid - my only concern with that is it's more difficult to reason about how configuration of global state impacts the behavior of the program. Ideally, I think it would be nice to have most of the logic be agnostic to the specific serialization mode.